File: slackoFrontend/slack/src/app/models/DTOs/BookDTO.ts
import {Author} from '../Author';
import {BookCategory} from '../BookCategory';

export class BookDTO {
  isbn: string;
  title: string;
  description: string;
  coverPictureUrl: string;
  publicationDate: Date;
  authors: Author[];
  categories: BookCategory[];

  constructor(isbn: string, title: string, description: string, coverPictureUrl: string,
              publicationDate: Date, authors: Author[], categories: BookCategory[]) {
    this.isbn = isbn;
    this.title = title;
    this.description = description;
    this.coverPictureUrl = coverPictureUrl;
    this.publicationDate = publicationDate;
    this.authors = authors;
    this.categories = categories;
  }
}

File: slackoBackend/src/main/java/com/slack/DTOs/BookDTO.java
package com.slack.DTOs;

import com.slack.entities.Author;
import com.slack.entities.BookCategory;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.validation.Valid;
import javax.validation.constraints.*;
import java.util.Date;
import java.util.List;

@Data
@AllArgsConstructor
@NoArgsConstructor
public class BookDTO {

    @Size(min = 10, max = 17, message = "ISBN must be 10-17 numbers long")
    @NotBlank(message = "ISBN is required")
    @Pattern(regexp = "(([0-9Xx][- ]*){13}|([0-9Xx][- ]*){10})",
            message = "ISBN is not valid")
    private String isbn;

    @Size(max = 1000, message = "Title can't exceed 1000 characters")
    @NotBlank(message = "Title is required")
    private String title;

    @Size(max = 2000, message = "Description can't exceed 2000 characters")
    private String description;

    @Size(max = 1000, message = "Cover picture URL can't exceed 1000 characters")
    private String coverPictureUrl;

    @Past(message = "The publication date of the book should be a past date")
    private Date publicationDate;

    @Valid
    @NotNull(message = "At least one book author is required.")
    private List<Author> authors;

    @Valid
    @NotNull(message = "At least one book category is required.")
    private List<BookCategory> categories;
}
